src/lib/Lower.hs

```diff
@@ -29,7 +33,325 @@ import GenericTraversal
+      _ -> error "unexpected decl emissions!"
+    _ -> error "unexpected emissions"
+{-# INLINE emitIx #-}
+
+--- The interesting instances ---
+
+class HasEmitIx (e::E) where
+  -- Ideally, we'd use an EmitsDicts predicate, kind of how we use EmitsInf within Inference.
+  -- But for now we just go with the slightly unsafe path and dynamic checks that verify
+  -- no unexpected emissions took place.
+  emitIxE :: Mut o => e i -> EmitIxM i o (e o)
+  default emitIxE :: (GenericE e, HasEmitIx (RepE e), Mut o) => e i -> EmitIxM i o (e o)
+  emitIxE e = confuseGHC >>= \_ -> toE <$> emitIxE (fromE e)
+
+class HasEmitIxB (b::B) where
+  emitIxB :: Mut o => b i i' -> (forall o'. DExt o o' => b o o' -> EmitIxM i' o' a) -> EmitIxM i o a
+
+-- DictExprs should appear in the IR only through DictCon, which we handle below.
+instance HasEmitIx DictExpr where
+  emitIxE _ = error "unhandled dict!"
+instance HasEmitIx Atom where
+  emitIxE = \case
+    DictCon de -> do
+      de' <- emitIxDict de
+      dictTy' <- getType de'
+      case dictTy' of
+        DictTy (DictType "Ix" _ _) -> do
+          EmitIxM $ SubstReaderT $ lift $ freshExtendSubInplaceT noHint \b ->
+            (EmitIxDictEmission $ Let b $ DeclBinding PlainLet dictTy' $ Atom $ DictCon de', Var $ binderName b)
+        -- Non-Ix dicts can be embedded in TypeCon params
+        _ -> return $ DictCon de'
+    atom -> toE <$> emitIxE (fromE atom)
+    where emitIxDict dict = toE <$> emitIxE (fromE dict)
+
+instance HasEmitIx Block where
+  emitIxE (Block _ idecls (ians :: Atom i')) = buildEmitIxBlock $ go idecls
+    where
+      go :: Mut o => Nest Decl i i' -> EmitIxM i o (Atom o)
+      go = \case
+        Empty -> emitIxE ians
+        Nest (Let b (DeclBinding ann _ e)) ds -> do
+          b' <- emitIxDecl (getNameHint b) ann =<< emitIxE e
+          extendSubst (b@>b') $ go ds
+
+-- TODO: This might unnecessarily hoist dicts above e.g. lambda binders.
+-- We could avoid doing that by specializing the LamExpr instance below.
+instance (BindsEnv b, SubstB Name b, SubstE Name e, SubstE AtomSubstVal e, SinkableE e, HasEmitIx e, HasEmitIxB b, BindsOneAtomName b) => HasEmitIx (Abs b e) where
+  emitIxE (Abs b body) = do
+    reemitIxDicts =<< emitIxB b \b' -> do
+      -- This scope is an overapproximation of the scope we need, because it also
+      -- includes b. It's unclear to me how to get rid of those casts, since we do
+      -- want the scope after all emissions coming from b's annotations, but before
+      -- b itself. And there's no easy way to get it with the way the code is set up
+      -- right now.
+      scope <- unsafeGetScope
+      bodyEms <- emitIxScoped $ emitIxE body
+      -- We only emit decls while traversing Blocks, and the implementation for
+      -- emitIxE should handle that locally. No other place should emit!
+      refreshAbs bodyEms \(EmitIxEmissions ds REmpty) body' -> do
+        (PairB ds' b'', s, _) <- return $ generalizeIxDicts (unsafeCoerceE scope) b' ds
```

Comment:
@dougalm please let me know if you have any suggestions about this use of unsafe APIs.